File: springapp/src/main/java/com/example/springapp/service/AppraisalService.java
package com.example.springapp.service;

import java.util.List;
import java.util.Optional;

import org.springframework.stereotype.Service;

import com.example.springapp.model.Appraisal;
import com.example.springapp.repository.AppraisalRepository;

@Service
public class AppraisalService {

    private final AppraisalRepository repo;

    public AppraisalService(AppraisalRepository repo) {
        this.repo = repo;
    }

    public List<Appraisal> getAllAppraisals() {
        return repo.findAll();
    }

    public Optional<Appraisal> getAppraisalById(Long id) {
        return repo.findById(id);
    }

    public List<Appraisal> getAppraisalsByEmployeeId(Long employeeId) {
        return repo.findByEmployee_EmployeeProfileId(employeeId);
    }

    public List<Appraisal> getAppraisalsByCycleId(Long cycleId) {
        return repo.findByReviewCycle_CycleId(cycleId);
    }

    public Appraisal createAppraisal(Appraisal appraisal) {
        if (appraisal.getEmployee() == null) {
            throw new IllegalArgumentException("Employee must be set for Appraisal");
        }
        
        // Set default values for new fields if not provided
        if (appraisal.getSelfRating() == null) {
            appraisal.setSelfRating(0);
        }
        if (appraisal.getManagerRating() == null) {
            appraisal.setManagerRating(0);
        }
        if (appraisal.getStatus() == null || appraisal.getStatus().isEmpty()) {
            appraisal.setStatus("Submitted");
        }
        
        return repo.save(appraisal);
    }

    public Appraisal updateAppraisal(Appraisal appraisal) {
        return repo.save(appraisal);
    }

    public void deleteAppraisal(Long id) {
        repo.deleteById(id);
    }
}

File: src/app/service/appraisal.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Appraisal } from '../model/appraisal.model';

@Injectable({
  providedIn: 'root'
})
export class AppraisalService {
  private apiUrl = 'http://localhost:8080/api/appraisals';

  constructor(private http: HttpClient) { }

  getAllAppraisals(): Observable<Appraisal[]> {
    return this.http.get<Appraisal[]>(this.apiUrl);
  }

  getAppraisalById(id: number): Observable<Appraisal> {
    return this.http.get<Appraisal>(`${this.apiUrl}/${id}`);
  }

  getAppraisalsByEmployeeId(employeeId: number): Observable<Appraisal[]> {
    return this.http.get<Appraisal[]>(`${this.apiUrl}/employee/${employeeId}`);
  }

  getAppraisalsByCycleId(cycleId: number): Observable<Appraisal[]> {
    return this.http.get<Appraisal[]>(`${this.apiUrl}/cycle/${cycleId}`);
  }

  createAppraisal(appraisal: Appraisal): Observable<Appraisal> {
    return this.http.post<Appraisal>(this.apiUrl, appraisal);
  }

  updateAppraisal(appraisal: Appraisal): Observable<Appraisal> {
    return this.http.put<Appraisal>(`${this.apiUrl}/${appraisal.appraisalId}`, appraisal);
  }

  deleteAppraisal(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
